### ISSUE TITLE
Adding support for a new programming language

### ISSUE THREAD

**Issue by Robert-Aron293** [googleapis/gapic-generator#3276]
I am trying to add support for D language and I don't know where to start.
What are the steps that I should follow?

**Comment by hkdevandla** [comment]
@Robert-Aron293  , can you clarify what exactly are you trying to request here. This code in this repo is used for generating Java client libraries. So not sure what exact support that you want to add. Please clarify.

**Comment by Robert-Aron293** [comment]
@hkdevandla i thought this code is used for generating client libraries for more than 1 language.
I wanted to know how the code generation works, so I can write a generator for another language (in my case D). Can you help me with that? Thanks!

**Comment by software-dov** [comment]
I strongly, strongly recommend that you don't use this codebase as the basis for a generator for a new language. It is in maintenance mode now, and some of the interfaces and datastructures are overly broad and general in an effort to support languages with highly different surfaces and semantics. Current best practice is to write a generator per target language as a protoc plugin.

[This](https://google.aip.dev/client-libraries/4210) link describes current best practices in more depth if you're interested in there being a generator for a new language. [aip.dev](https://google.aip.dev) is a great resource for best practices and requirements when writing a new generator. Let us know if you have any questions!

**Comment by Robert-Aron293** [comment]
Hello @software-dov !
I am currently working on "D Language Client Libraries for Google APIs" project[0][1].
I have 2 questions:
The only thing that I have to do is to write a protoc plugin?
If so, can you explain how the code generation works or where can I find some documentation on that?

Thanks!

[0] https://github.com/dlang/projects/issues/66
[1]https://forum.dlang.org/post/eaefvlbgikujonfjjzqi@forum.dlang.org

**Comment by software-dov** [comment]
Correct, writing a protoc plugin is all that is needed. Protoc itself is responsible for parsing and serializing input proto files *and* deserializing the generated client library surface.

This [post](https://expobrain.net/2015/09/13/create-a-plugin-for-google-protocol-buffer/) is a good introduction into how protoc plugins work and how they interact with protoc. The example is in python, but the important parts are generalizable.

One complication is D support for protobuf messages. You may find it easier to write the plugin in a different language with more protobuf documentation, but you know your tools and workflow best.

Take a look at other language generators for [Python](https://github.com/googleapis/gapic-generator-python/), [golang](https://github.com/googleapis/gapic-generator-go), and [Java](https://github.com/googleapis/gapic-generator-java) if you want to see how other languages set up templates, interact with protobuf descriptors, and structure internal data.

**Comment by Robert-Aron293** [comment]
@software-dov thanks for the help! Have you tried to generate client libraries for google drive api, for example? If so, how can I do it? I can't really find the protos associated with this api.

**Comment by software-dov** [comment]
To the best of my knowledge neither this generator nor any of the generators listed above have ever been used to create client libraries for Drive. I'm not sure where the Drive protos are; they don't appear to be visible within [googleapis/googleapis](https://github.com/googleapis/googleapis), which has the proto definition for most of the other APIs.

**Comment by Robert-Aron293** [comment]
But do those Drive protos exists? If not, should I write them or should I write the client library for GDrive API?
